src/box/port.c

```diff
@@ -98,143 +165,205 @@ port_c_new_entry(struct port_c *port)
+	 * address of the entry for this purpose.
+	 */
+	if (len == 0) {
+		str = (char *)pe;
+	} else {
+		void *ptr = port_c_data_xalloc(len);
+		memcpy(ptr, data, len);
+		str = ptr;
 	}
-	if (size <= PORT_ENTRY_SIZE)
-		mempool_free(&port_entry_pool, dst);
-	else
-		free(dst);
-	return NULL;
+	pe->value.data.data = str;
+	pe->value.data.size = len;
 }
 
-int
-port_c_add_mp(struct port *base, const char *mp, const char *mp_end)
+void
+port_c_add_null(struct port *base)
 {
-	assert(mp_end > mp);
-	uint32_t mp_size = mp_end - mp;
-	char *dst = port_c_prepare_mp(base, mp_size);
-	if (dst == NULL)
-		return -1;
-	memcpy(dst, mp, mp_end - mp);
-	return 0;
+	struct port_c *port = (struct port_c *)base;
+	struct port_c_entry *pe = port_c_new_entry(port);
+	pe->type = PORT_C_ENTRY_NULL;
 }
 
-int
-port_c_add_formatted_mp(struct port *base, const char *mp, const char *mp_end,
-			struct tuple_format *format)
+void
+port_c_add_bool(struct port *base, bool val)
 {
-	int rc = port_c_add_mp(base, mp, mp_end);
-	if (rc != 0)
-		return rc;
 	struct port_c *port = (struct port_c *)base;
-	port->last->mp_format = format;
-	tuple_format_ref(format);
-	return 0;
+	struct port_c_entry *pe = port_c_new_entry(port);
+	pe->type = PORT_C_ENTRY_BOOL;
+	pe->value.boolean = val;
 }
 
-int
-port_c_add_str(struct port *base, const char *str, uint32_t len)
+void
+port_c_add_double(struct port *base, double val)
 {
-	uint32_t mp_size = mp_sizeof_str(len);
-	char *dst = port_c_prepare_mp(base, mp_size);
-	if (dst == NULL)
-		return -1;
-	mp_encode_str(dst, str, len);
-	return 0;
+	struct port_c *port = (struct port_c *)base;
+	struct port_c_entry *pe = port_c_new_entry(port);
+	pe->type = PORT_C_ENTRY_DOUBLE;
+	pe->value.number = val;
+}
+
+void
+port_c_add_iterator(struct port *base, void *data,
+		    port_c_iterator_create_f create,
+		    port_c_iterator_next_f next,
+		    port_c_iterator_destroy_f destroy)
+{
+	static_assert(sizeof(struct port_c_iterator) <= PORT_ENTRY_SIZE,
+		      "Iterator must fit port_c_entry to use its mempool");
+
+	struct port_c *port = (struct port_c *)base;
+	struct port_c_entry *pe = port_c_new_entry(port);
+	pe->type = PORT_C_ENTRY_ITER;
+	pe->value.iter = xmempool_alloc(&port_entry_pool);
+	pe->value.iter->data = data;
+	pe->value.iter->create = create;
+	pe->value.iter->next = next;
+	pe->value.iter->destroy = destroy;
+}
+
+/** Callback to forward and error from mpstream methods. */
+static inline void
+set_encode_error(void *error_ctx)
+{
+	*(bool *)error_ctx = true;
 }
 
 static int
 port_c_dump_msgpack(struct port *base, struct obuf *out, struct mp_ctx *ctx)
 {
 	struct port_c *port = (struct port_c *)base;
 	struct port_c_entry *pe;
-	for (pe = port->first; pe != NULL; pe = pe->next) {
-		uint32_t size = pe->mp_size;
-		if (size == 0) {
+	struct mpstream stream;
+	bool obuf_failed = false;
+	mpstream_init(&stream, out, obuf_reserve_cb, obuf_alloc_cb,
+		      set_encode_error, &obuf_failed);
```

Comment:
Let's use `mpstream_panic_cb` here for simplicity. We don't expect OOM from obuf allocations.

Reply:
Done.